### PR TITLE
openFOAM-paraview-catalyst

### PR DESCRIPTION
**Pull request by sbna**
Merged PR #8752 and #8753 in a unique PR as we agreed. PR #8752 and #8753 can be closed.
Updated openfoam, paraview and catalyst packages and added the new package of-catalyst
to couple openfoam with catalyst to perform in situ visualization.

### CODE REVIEW

File: var/spack/repos/builtin/packages/of-catalyst/package.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    This offering is part of the community repository supported by OpenCFD Ltd,
+    producer and distributor of the OpenFOAM software via www.openfoam.com,
+    and owner of the OPENFOAM trademark.
+    OpenCFD Ltd has been developing and releasing OpenFOAM since its debut
+    in 2004.
+    """
+
+    # Currently only via git
+    homepage = "https://develop.openfoam.com/Community/catalyst"
+    gitrepo  = "https://develop.openfoam.com/Community/catalyst.git"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/of-catalyst/package.py]: If you name this `git`, Spack now supports package-level git tags, so you won't need to specify it in `version()` below.

File: var/spack/repos/builtin/packages/of-catalyst/package.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    version('1806', git=gitrepo, tag='v1806')
+    version('develop', branch='develop', git=gitrepo)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/of-catalyst/package.py]: I would put develop above 1806 as it's newer.

File: var/spack/repos/builtin/packages/of-catalyst/package.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    variant('full', default=False, description='Build against paraview (full) or catalyst (light)')
+
+    depends_on('openfoam-com@1806', when='@1806', type=('build', 'link', 'run'))
+    depends_on('openfoam-com@develop', when='@develop', type=('build', 'link', 'run'))
+    depends_on('catalyst@5.5:', when='~full')
+    depends_on('paraview@5.5:+osmesa~qt', when='+full')
+
+    def setup_environment(self, spack_env, run_env):
+        run_env.prepend_path('LD_LIBRARY_PATH', join_path(self.prefix,
+                             'lib'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/of-catalyst/package.py]: This should automatically be in your module file, so you shouldn't need this.

File: var/spack/repos/builtin/packages/of-catalyst/package.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    @property
+    def root_cmakelists_dir(self):
+        """The relative path to the directory containing CMakeLists.txt
+
+        This path is relative to the root of the extracted tarball,
+        not to the ``build_directory``. Defaults to the current directory.
+
+        :return: directory containing CMakeLists.txt
+        """
+        return join_path(self.stage.source_path, join_path('src', 'catalyst'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/of-catalyst/package.py]: You can also do:
```python
root_cmakelists_dir = 'src/catalyst'
```

File: var/spack/repos/builtin/packages/of-catalyst/package.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    def cmake_args(self):
+        """Populate cmake arguments for ParaView."""
+        cmake_args = [
+            '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY:PATH=%s' % join_path(
+                self.stage.source_path,
+                'spack-build'),
+            '-DCMAKE_INSTALL_PREFIX:PATH=%s' % self.prefix
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/of-catalyst/package.py]: This arg isn't needed, it's added by the `CMakePackage` base class already.

File: var/spack/repos/builtin/packages/catalyst/package.py [hunk 6]
```diff
@@ -104,15 +120,29 @@ def do_stage(self, mirror_only=False):
                                                     self.stage.path))
 
     def setup_environment(self, spack_env, run_env):
+        # paraview 5.5 and later
+        # - cmake under lib/cmake/paraview-5.5
+        # - libs  under lib
+        # - python bits under lib/python2.8/site-packages
```
Comment (jrood-nrel) [inline var/spack/repos/builtin/packages/catalyst/package.py]: I think this should be `lib/python2.7/site-packages`

File: var/spack/repos/builtin/packages/of-catalyst/package.py (hunk 4 above)
Comment (jrood-nrel) [reply var/spack/repos/builtin/packages/of-catalyst/package.py]: To be more explicit, I have an example of this here https://github.com/Exawind/build-test/blob/master/configs/custom-package-files/trilinos-catalyst-ioss-adapter/package.py

File: var/spack/repos/builtin/packages/of-catalyst/package.py (hunk 5 above)
Comment (sbna) [reply var/spack/repos/builtin/packages/of-catalyst/package.py]: yes, -DCMAKE_INSTALL_PREFIX is already there, line 144, but -DCMAKE_LIBRARY_OUTPUT_DIRECTORY not. Can I remove -DCMAKE_INSTALL_PREFIX and leave -DCMAKE_LIBRARY_OUTPUT_DIRECTORY.

File: var/spack/repos/builtin/packages/of-catalyst/package.py (hunk 1 above)
Comment (sbna) [reply var/spack/repos/builtin/packages/of-catalyst/package.py]: Sorry, name what with git?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/of-catalyst/package.py]:
```suggestion
    git  = "https://develop.openfoam.com/Community/catalyst.git"
```

File: var/spack/repos/builtin/packages/of-catalyst/package.py (hunk 5 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/of-catalyst/package.py]: Yep, I was only talking about `-DCMAKE_INSTALL_PREFIX`.

File: var/spack/repos/builtin/packages/of-catalyst/package.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    @property
+    def root_cmakelists_dir(self):
+        """The relative path to the directory containing CMakeLists.txt
+
+        This path is relative to the root of the extracted tarball,
+        not to the ``build_directory``. Defaults to the current directory.
+
+        :return: directory containing CMakeLists.txt
+        """
+        root_cmakelists_dir = 'src/catalyst'
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/of-catalyst/package.py]: Sorry for not being more clear, you can replace this entire function with the line above, so the function is no longer needed.

File: var/spack/repos/builtin/packages/of-catalyst/package.py (hunk 1 above)
Comment (sbna) [reply var/spack/repos/builtin/packages/of-catalyst/package.py]: done

File: var/spack/repos/builtin/packages/of-catalyst/package.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    and owner of the OPENFOAM trademark.
+    OpenCFD Ltd has been developing and releasing OpenFOAM since its debut
+    in 2004.
+    """
+
+    # Currently only via git
+    homepage = "https://develop.openfoam.com/Community/catalyst"
+    git = "https://develop.openfoam.com/Community/catalyst.git"
+
+    version('develop', branch='develop', git=gitrepo)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/of-catalyst/package.py]: You need to drop `git=gitrepo` now.

Comment (sbna) [reply var/spack/repos/builtin/packages/of-catalyst/package.py]: sorry

File: var/spack/repos/builtin/packages/of-catalyst/package.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    OpenCFD Ltd has been developing and releasing OpenFOAM since its debut
+    in 2004.
+    """
+
+    # Currently only via git
+    homepage = "https://develop.openfoam.com/Community/catalyst"
+    git = "https://develop.openfoam.com/Community/catalyst.git"
+
+    version('develop', branch='develop')
+    version('1806', git=gitrepo, tag='v1806')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/of-catalyst/package.py]: Same here.

Comment (sbna) [reply var/spack/repos/builtin/packages/of-catalyst/package.py]: sorry again

**Comment by sbna** [comment]
I don't see any errors this time, why Travis failed?

**Pull request merged**